File: java/sfo_06_print_a_linked_list_in_reverse_order_s1/sfo_06_print_a_linked_list_in_reverse_order_s1.java
/*
* File: sfo_06_print_a_linked_list_in_reverse_order_s1.java
* Created Time: 2021-12-09
* Author: Krahets (krahets@163.com)
*/

package sfo_06_print_a_linked_list_in_reverse_order_s1;

import include.*;
import java.util.*;

// ===== Solution Code =====
class Solution {
    ArrayList<Integer> tmp = new ArrayList<Integer>();
    public int[] reversePrint(ListNode head) {
        recur(head);
        int[] res = new int[tmp.size()];
        for(int i = 0; i < res.length; i++)
            res[i] = tmp.get(i);
        return res;
    }
    void recur(ListNode head) {
        if(head == null) return;
        recur(head.next);
        tmp.add(head.val);
    }
}

public class sfo_06_print_a_linked_list_in_reverse_order_s1 {
    public static void main(String[] args) {
        // ======= Test Case =======
        ListNode head = ListNode.arrToLinkedList(new int[] { 1, 3, 2 });
        // ====== Driver Code ======
        Solution slt = new Solution();
        int[] res = slt.reversePrint(head);
        System.out.println(Arrays.toString(res));
    }
}

File: sword_for_offer/codes/python/sfo_06_print_a_linked_list_in_reverse_order_s1.py
'''
File: sfo_06_print_a_linked_list_in_reverse_order_s1.py
Created Time: 2021-12-09
Author: Krahets (krahets@163.com)
'''

from include import *

# ===== Solution Code =====
class Solution:
    def reversePrint(self, head: ListNode) -> List[int]:
        return self.reversePrint(head.next) + [head.val] if head else []

# ======= Test Case =======
head = list_to_linked_list([1, 3, 2])
# ====== Driver Code ======
slt = Solution()
res = slt.reversePrint(head)
print(res)
